cmds/init/init.go

```diff
@@ -31,6 +33,95 @@ var (
 	debug   = func(string, ...interface{}) {}
 )
 
+// fillBuildBin tries to fill buildbin. It does it's level best but if it can not do it,
+// c'est la vie.
+func fillBuildBin() {
+	var (
+		// mark that we visited a directory, remember the name
+		// of the go file in that directory. If the name is "",
+		// we won't bother to test it.
+		visited = map[string]string{
+			"/src/github.com/u-root/u-root/cmds/installcommand": "",
+			"/src/github.com/u-root/u-root/cmds/init":           "",
+			"/src/github.com/u-root/u-root/cmds/rush":           "",
+		}
+		// populate is the list of things we must populate.
+		// If you know something you just gotta have, then put it here.
+		populate = []string{"rush"}
+	)
+
+	// populate buildbin
+	f, err := find.New(func(f *find.Finder) error {
+		f.Pattern = "*.go"
+		f.Root = "/src"
+		return nil
+	})
+	if err != nil {
+		log.Printf("Can't build a find, sorry, err %v", err)
+		return
+	}
+	go f.Find()
+
+	for goFile := range f.Names {
+		if goFile.Err != nil {
+			log.Printf("%s: got %v, want nil", goFile.Name, goFile.Err)
+			continue
+		}
+		dir := filepath.Dir(goFile.Name)
+		if _, ok := visited[dir]; ok {
+			continue
+		}
+		visited[dir] = goFile.Name
+	}
+	pm, err := regexp.Compile("\npackage  *main")
+	if err != nil {
+		log.Printf("package main regexp did not compile; finding non-uroot packages won't work")
+	}
+
+	for d, f := range visited {
+		if f == "" {
+			continue
+		}
+		// The simple common case is a base u-root command,
+		// which is easy: the Dir of the dir is
+		// /src/github.com/u-root/u-root/cmds. Check that first.
+		if filepath.Dir(d) == "/src/github.com/u-root/u-root/cmds" {
+			log.Printf("Fast path for %v", d)
+			populate = append(populate, filepath.Base(d))
+			continue
+		}
+		if pm == nil {
+			continue
+		}
+		// We should, really, use ast and look for the package.
+		// But that's kinda slower, so we'll do this hack until
+		// Chris and Ryan force it to be done right.
+		// I'm not yet convinced that just sucking the whole file
+		// in is significantly slower than using MatchReader.
+		// We've already gained a lot with the simple fast path rule
+		// above, so this is likely OK.
+		b, err := ioutil.ReadFile(f)
+		if err != nil {
+			log.Printf("Reading %v: got %v, want nil", f, err)
+			continue
+		}
+		if !pm.Match(b) {
```

Comment:
Doesn't this mean that any file that isn't package main will get included in here if the file contains a string or something that matches? One example is pkg/uroot/bb.go. I understand that that may never be included in the ramfs mode anyway, and even if you generate the symlink that may not be a problem?

Reply:
I don't think this is a blocker, I just think maybe you should make it explicit in a comment somewhere so people don't get tripped up, if ever.

Reply:
note the string is this:
\npackage . *main
i.e. the package has to be the first thing on the line, has to have only spaces, followed by main. 
Yep, if there is a long utf-8 and that's on a line, this will create a bogus symlink.

A quick git grep '^package . *main'
(note: package<SPACE><SPACE>*main', stupid editor changed two spaces ' . ')

finds no false postives. So this admitted hack saves us a full parse of all those files. 

I would argue this is a hack that works, I can add a comment to that effect.

Reply:
Packages don't always have "package main" on the first line. There can be a package comment.

I don't think you can get around at least doing some parsing.

Reply:
This works whether it is on the first line or not as far as I can tell. But let me try.
yep, matches if there is a \n followed by package . *main
anywhere in the file.

But if you want to show me a simple parsing thing, I'm all ears, as long as it doesn't blow up the startup time. by more than a bit.

Reply:
https://play.golang.org/p/KP8tgi9QA_

Reply:
something weird going on with the shared thing, there should be a URL there but github seems to be stripping it our for me.
play.golang.org/p/KP8tgi9QA_

Reply:
[https://play.golang.org/p/KP8tgi9QA_](https://play.golang.org/p/KP8tgi9QA_)

Reply:
You could also use regex FindAll on `^package`, then check if the first result matches `package +main`